### PR TITLE
[FLINK-29480][Connectors/Kafka] Skip null records when writing

### PR DESCRIPTION
**Pull request by salvalcantara**
## What is the purpose of the change

Make `KafkaWriter` skip null records since sending them is meaningless in the first place. The proposed behaviour has two main advantages:
1. It prevents a `NullPointerException`
2. It is more consistent with the _reading case_, where one can skip the null records returned by the deserializer

## Brief change log

Add simple if statement in `KafkaWriter#serialize` for skipping them.

## Verifying this change

This change is already covered by the existing test `KafkaWriterITCase#testIncreasingRecordBasedCounters`, which has been slightly modified to account for null records.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

**Comment by MartijnVisser** [comment]
@salvalcantara There's an incompatibility in your patch that needs to be addressed. 

```
[ERROR] Failed to execute goal io.github.zentol.japicmp:japicmp-maven-plugin:0.16.0_m325:cmp (default) on project flink-core: There is at least one incompatibility: org.apache.flink.api.connector.source.SourceReader.pauseOrResumeSplits(java.util.Collection,java.util.Collection):METHOD_NEW_DEFAULT -> [Help 1]
```

**Comment by MartijnVisser** [comment]
Also keep in mind that this patch most likely also needs a test where this functionality is indeed tested

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/sink/KafkaWriterITCase.java [hunk 1]
```diff
@@ -491,6 +498,9 @@ private class DummyRecordSerializer implements KafkaRecordSerializationSchema<In
         @Override
         public ProducerRecord<byte[], byte[]> serialize(
                 Integer element, KafkaSinkContext context, Long timestamp) {
+            if (element == null) {
+                return null;
+            }
             return new ProducerRecord<>(topic, ByteBuffer.allocate(4).putInt(element).array());
         }
     }
```
Comment (salvalcantara) [inline flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/sink/KafkaWriterITCase.java:506]: Isn't this update of the current `testIncreasingRecordBasedCounters` case enough for testing the change @MartijnVisser?

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/sink/KafkaWriterITCase.java [hunk 2]
```diff
@@ -148,6 +148,13 @@ public void testIncreasingRecordBasedCounters() throws Exception {
             assertThat(numRecordsOutErrors.getCount()).isEqualTo(0);
             assertThat(numRecordsSendErrors.getCount()).isEqualTo(0);
 
+            // A null record is meaningless and should be skipped
+            writer.write(null, SINK_WRITER_CONTEXT);
+            assertThat(numBytesOut.getCount()).isEqualTo(0L);
+            assertThat(numRecordsOut.getCount()).isEqualTo(0);
+            assertThat(numRecordsOutErrors.getCount()).isEqualTo(0);
+            assertThat(numRecordsSendErrors.getCount()).isEqualTo(0);
+
```
Comment (salvalcantara) [inline flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/sink/KafkaWriterITCase.java:157]: Isn't this enough @MartijnVisser for testing the changes? I can consider a separate test for the null records exclusively but the existing `testIncreasingRecordBasedCounters` looks like a good spot to me.

**Comment by salvalcantara** [comment]
@flinkbot run azure

**Comment by salvalcantara** [comment]
Initially, the ci pipeline passed successfully, I made some cosmetic changes and then I got a failure which seems unrelated to my code changes, really. Can you confirm @MartijnVisser @mas-chen @leonardBang @PatrickRen?

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/sink/KafkaWriterITCase.java [hunk 3]
```diff
@@ -148,6 +148,15 @@ public void testIncreasingRecordBasedCounters() throws Exception {
             assertThat(numRecordsOutErrors.getCount()).isEqualTo(0);
             assertThat(numRecordsSendErrors.getCount()).isEqualTo(0);
 
+            // elements that cannot be serialized should be silently skipped
```
Comment (leonardBang) [inline flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/sink/KafkaWriterITCase.java:151]: minor: we can improve the note, because not all serializers will return `null` when the element cannot be serialized, not all `null` are due to cannot be serialized . 
What we can ensure is only that the `KafkaSinkWriter` will silently skip `null` that returned by serializers

**Review by leonardBang — COMMENTED** [review]
Thanks @salvalcantara for the contribution, I left some comments about detail.

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/sink/KafkaWriterITCase.java (hunk 3 above)
Comment (salvalcantara) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/sink/KafkaWriterITCase.java:151]: Agreed. Maybe something along these lines?

```suggestion
            // elements for which the serializer returns null should be silently skipped
```

Comment (leonardBang) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/sink/KafkaWriterITCase.java:151]: > Agreed. Maybe something along these lines?

Yes, this is better.

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/sink/KafkaWriterITCase.java [hunk 4]
```diff
@@ -148,6 +148,15 @@ public void testIncreasingRecordBasedCounters() throws Exception {
...
+            writer.write(null, SINK_WRITER_CONTEXT);
+            timeService.trigger();
+            assertThat(numBytesOut.getCount()).isEqualTo(0L);
+            assertThat(numRecordsOut.getCount()).isEqualTo(0);
+            assertThat(numRecordsOutErrors.getCount()).isEqualTo(0);
+            assertThat(numRecordsSendErrors.getCount()).isEqualTo(0);
+
+            // but properly serialized elements should count just normally
```
Comment (salvalcantara) [inline flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/sink/KafkaWriterITCase.java:159]: @leonardBang I would also update this companion note
```suggestion
            // but elements for which a non-null producer record is returned should count
```

[the remark this answers is not in the document]
Comment (salvalcantara) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/sink/KafkaWriterITCase.java:159]: I considered that too, indeed that was my initial plan:

- Add another dummy serializer, e.g., `NullRecordSerializer`, which simply returns `null` regardless of the input element
- Add a separate test, e.g., `testWriteNullProducerRecord`, along the lines of your suggestion

In the end, I didn't do so because it seems simpler and at the same time more realistic to me having the current test (`testIncreasingRecordBasedCounters`) check that we get the counting right when interleaving both valid and invalid elements (using the same `NullRecordSerializer`), as opposed to doing so in isolation. Also note that new separate test would essentially be a clone of the existing one, plus require yet another`createWriterWithConfiguration` method/override that considers the new dummy serializer instead.

In summary, I think `testIncreasingRecordBasedCounters` is actually a good spot for checking the new functionality. Having said that, we can of course create a separate test if my arguments are not convincing enough.

[the remark this answers is not in the document]
Comment (leonardBang) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/sink/KafkaWriterITCase.java:159]: Thanks @salvalcantara for the detail explanation, Adding to current test could check the metrics and checks the writing null ProduceRecord as well, it makes sense to me.

**Review by leonardBang — APPROVED** [review]
Thanks @salvalcantara for the update, LGTM. 
Let's wait the CI passed.

**Comment by MartijnVisser** [comment]
@salvalcantara Please run a `mvn spotless:apply` and push the fixes to get the necessary styling in (which currently fails the CI)

**Comment by salvalcantara** [comment]
@MartijnVisser @leonardBang The CI finally passed, so we should be ready to go.
